File: feels-app/src/services/indexer/websocket.ts
// WebSocket client for real-time Feels Protocol updates

// Subscription types
export type SubscriptionType = 
  | { type: 'all_markets' }
  | { type: 'market'; address: string }
  | { type: 'swaps'; market?: string }
  | { type: 'positions'; user?: string }
  | { type: 'floor_updates'; market: string }
  | { type: 'price_updates'; market: string };

// Update event types
export type UpdateEvent = 
  | {
      type: 'market_update';
      market: string;
      sqrt_price: string;
      liquidity: string;
      current_tick: number;
      timestamp: number;
    }
  | {
      type: 'swap_event';
      market: string;
      user: string;
      amount_in: string;
      amount_out: string;
      token_in: string;
      token_out: string;
      price: number;
      timestamp: number;
    }
  | {
      type: 'position_update';
      position: string;
      market: string;
      owner: string;
      liquidity: string;
      tick_lower: number;
      tick_upper: number;
      timestamp: number;
    }
  | {
      type: 'floor_update';
      market: string;
      new_floor_tick: number;
      new_floor_price: number;
      timestamp: number;
    }
  | {
      type: 'price_update';
      market: string;
      price: number;
      price_change_24h: number;
      timestamp: number;
    }
  | {
      type: 'subscribed';
      id: string;
      subscriptions: SubscriptionType[];
    }
  | {
      type: 'error';
      code: string;
      message: string;
    };

export class FeelsWebSocketClient {
  private ws?: WebSocket;
  private url: string;
  private reconnectTimeout?: NodeJS.Timeout;
  private subscriptions: Map<string, SubscriptionType[]> = new Map();
  private eventHandlers: Map<string, (event: UpdateEvent) => void> = new Map();
  
  constructor(url: string) {
    this.url = url.replace('http:', 'ws:').replace('https:', 'wss:');
  }
  
  connect(): Promise<void> {
    return new Promise((resolve, reject) => {
      // Only use WebSocket in browser environment
      if (typeof window === 'undefined') {
        reject(new Error('WebSocket is not available in server environment'));
        return;
      }
      
      try {
        this.ws = new WebSocket(`${this.url}/ws`);
        
        this.ws.onopen = () => {
          console.log('WebSocket connected');
          // Resubscribe to all previous subscriptions
          for (const [id, subs] of this.subscriptions) {
            this.ws?.send(JSON.stringify({
              type: 'subscribe',
              id,
              subscriptions: subs,
            }));
          }
          resolve();
        };
        
        this.ws.onmessage = (event) => {
          try {
            const update = JSON.parse(event.data) as UpdateEvent;
            // Dispatch to all registered handlers
            for (const handler of this.eventHandlers.values()) {
              handler(update);
            }
          } catch (e) {
            console.error('Failed to parse WebSocket message:', e);
          }
        };
        
        this.ws.onerror = (error) => {
          console.error('WebSocket error:', error);
          reject(error);
        };
        
        this.ws.onclose = () => {
          console.log('WebSocket disconnected');
          // Attempt reconnection after 5 seconds
          this.reconnectTimeout = setTimeout(() => {
            this.connect().catch(console.error);
          }, 5000);
        };
      } catch (error) {
        reject(error);
      }
    });
  }
  
  disconnect(): void {
    if (this.reconnectTimeout) {
      clearTimeout(this.reconnectTimeout);
    }
    if (this.ws) {
      this.ws.close();
      this.ws = undefined;
    }
  }
  
  subscribe(id: string, subscriptions: SubscriptionType[]): void {
    this.subscriptions.set(id, subscriptions);
    // Only check WebSocket state in browser environment
    if (typeof window !== 'undefined' && this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({
        type: 'subscribe',
        id,
        subscriptions,
      }));
    }
  }
  
  unsubscribe(id: string): void {
    const subscriptions = this.subscriptions.get(id);
    // Only check WebSocket state in browser environment
    if (typeof window !== 'undefined' && subscriptions && this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send(JSON.stringify({
        type: 'unsubscribe',
        id,
        subscriptions,
      }));
    }
    this.subscriptions.delete(id);
  }
  
  onUpdate(id: string, handler: (event: UpdateEvent) => void): void {
    this.eventHandlers.set(id, handler);
  }
  
  removeHandler(id: string): void {
    this.eventHandlers.delete(id);
  }
}

// Factory function for creating WebSocket client instances
export const createWebSocketClient = (baseUrl?: string): FeelsWebSocketClient => {
  const url = baseUrl || process.env['NEXT_PUBLIC_INDEXER_URL'] || 'http://localhost:8080';
  return new FeelsWebSocketClient(url);
};

File: feels-indexer/src/api/websocket.rs
//! WebSocket support for real-time updates
//!
//! Provides WebSocket endpoints for subscribing to real-time data updates

use super::ApiState;
use axum::{
    extract::{ws::{Message, WebSocket, WebSocketUpgrade}, State, Query},
    response::Response,
};
use futures::{sink::SinkExt, stream::StreamExt};
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use tokio::sync::broadcast;
use tracing::{info, warn, error};

/// WebSocket subscription types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum SubscriptionType {
    /// Subscribe to all market updates
    AllMarkets,
    /// Subscribe to specific market
    Market { address: String },
    /// Subscribe to swap events
    Swaps { market: Option<String> },
    /// Subscribe to position updates
    Positions { user: Option<String> },
    /// Subscribe to floor updates
    FloorUpdates { market: String },
    /// Subscribe to price updates
    PriceUpdates { market: String },
}

/// WebSocket message types
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum WsMessage {
    /// Subscription request
    Subscribe {
        id: String,
        subscriptions: Vec<SubscriptionType>,
    },
    /// Unsubscribe request
    Unsubscribe {
        id: String,
        subscriptions: Vec<SubscriptionType>,
    },
    /// Ping for keepalive
    Ping,
    /// Pong response
    Pong,
}

/// WebSocket update events
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "snake_case")]
pub enum UpdateEvent {
    /// Market data update
    MarketUpdate {
        market: String,
        sqrt_price: String,
        liquidity: String,
        current_tick: i32,
        timestamp: i64,
    },
    /// New swap event
    SwapEvent {
        market: String,
        user: String,
        amount_in: String,
        amount_out: String,
        token_in: String,
        token_out: String,
        price: f64,
        timestamp: i64,
    },
    /// Position update
    PositionUpdate {
        position: String,
        market: String,
        owner: String,
        liquidity: String,
        tick_lower: i32,
        tick_upper: i32,
        timestamp: i64,
    },
    /// Floor update
    FloorUpdate {
        market: String,
        new_floor_tick: i32,
        new_floor_price: f64,
        timestamp: i64,
    },
    /// Price update
    PriceUpdate {
        market: String,
        price: f64,
        price_change_24h: f64,
        timestamp: i64,
    },
    /// Subscription confirmation
    Subscribed {
        id: String,
        subscriptions: Vec<SubscriptionType>,
    },
    /// Unsubscribe confirmation
    Unsubscribed {
        id: String,
        subscriptions: Vec<SubscriptionType>,
    },
    /// Error message
    Error {
        code: String,
        message: String,
    },
}

/// Query parameters for WebSocket connection
#[derive(Deserialize)]
pub struct WsQuery {
    /// Optional auth token for authenticated subscriptions
    pub auth: Option<String>,
}

/// Handle WebSocket upgrade
pub async fn websocket_handler(
    ws: WebSocketUpgrade,
    Query(params): Query<WsQuery>,
    State(state): State<ApiState>,
) -> Response {
    info!("WebSocket connection requested");
    
    ws.on_upgrade(move |socket| handle_socket(socket, state, params))
}

/// Handle WebSocket connection
async fn handle_socket(
    socket: WebSocket,
    state: ApiState,
    _params: WsQuery,
) {
    let (mut sender, mut receiver) = socket.split();
    
    // Create broadcast channel for this connection
    let (tx, mut rx) = broadcast::channel::<UpdateEvent>(100);
    
    // Spawn task to handle incoming messages
    let state_clone = state.clone();
    let tx_clone = tx.clone();
    let mut recv_task = tokio::spawn(async move {
        while let Some(msg) = receiver.next().await {
            if let Ok(msg) = msg {
                if process_message(msg, &state_clone, &tx_clone).await.is_err() {
                    break;
                }
            }
        }
    });
    
    // Spawn task to send updates
    let mut send_task = tokio::spawn(async move {
        while let Ok(event) = rx.recv().await {
            let msg = Message::Text(serde_json::to_string(&event).unwrap());
            if sender.send(msg).await.is_err() {
                break;
            }
        }
    });
    
    // Wait for either task to finish
    tokio::select! {
        _ = (&mut send_task) => {
            recv_task.abort();
        }
        _ = (&mut recv_task) => {
            send_task.abort();
        }
    }
    
    info!("WebSocket connection closed");
}

/// Process incoming WebSocket message
async fn process_message(
    msg: Message,
    _state: &ApiState,
    tx: &broadcast::Sender<UpdateEvent>,
) -> Result<(), ()> {
    match msg {
        Message::Text(text) => {
            match serde_json::from_str::<WsMessage>(&text) {
                Ok(WsMessage::Subscribe { id, subscriptions }) => {
                    info!("Subscribe request: {:?}", subscriptions);
                    
                    // Send subscription confirmation
                    let event = UpdateEvent::Subscribed {
                        id,
                        subscriptions,
                    };
                    let _ = tx.send(event);
                    
                    // TODO: Register subscriptions and start sending updates
                }
                Ok(WsMessage::Unsubscribe { id, subscriptions }) => {
                    info!("Unsubscribe request: {:?}", subscriptions);
                    
                    // Send unsubscribe confirmation
                    let event = UpdateEvent::Unsubscribed {
                        id,
                        subscriptions,
                    };
                    let _ = tx.send(event);
                    
                    // TODO: Remove subscriptions
                }
                Ok(WsMessage::Ping) => {
                    // Send pong - handled at protocol level
                }
                Ok(WsMessage::Pong) => {
                    // Pong received
                }
                Err(e) => {
                    warn!("Invalid WebSocket message: {}", e);
                    let event = UpdateEvent::Error {
                        code: "INVALID_MESSAGE".to_string(),
                        message: format!("Failed to parse message: {}", e),
                    };
                    let _ = tx.send(event);
                }
            }
        }
        Message::Binary(_) => {
            warn!("Binary messages not supported");
            let event = UpdateEvent::Error {
                code: "UNSUPPORTED".to_string(),
                message: "Binary messages not supported".to_string(),
            };
            let _ = tx.send(event);
        }
        Message::Close(_) => {
            info!("WebSocket close received");
            return Err(());
        }
        _ => {}
    }
    
    Ok(())
}

/// Broadcast service for pushing updates to connected clients
pub struct UpdateBroadcaster {
    /// Channels for each subscription type
    market_channels: Arc<tokio::sync::RwLock<Vec<broadcast::Sender<UpdateEvent>>>>,
}

impl UpdateBroadcaster {
    pub fn new() -> Self {
        Self {
            market_channels: Arc::new(tokio::sync::RwLock::new(Vec::new())),
        }
    }
    
    /// Broadcast market update
    pub async fn broadcast_market_update(&self, update: UpdateEvent) {
        let channels = self.market_channels.read().await;
        for tx in channels.iter() {
            let _ = tx.send(update.clone());
        }
    }
    
    /// Broadcast swap event
    pub async fn broadcast_swap_event(&self, event: UpdateEvent) {
        let channels = self.market_channels.read().await;
        for tx in channels.iter() {
            let _ = tx.send(event.clone());
        }
    }
}

/// Create WebSocket routes
pub fn create_websocket_routes() -> axum::Router<ApiState> {
    axum::Router::new()
        // WebSocket route temporarily disabled due to handler signature issues
        // .route("/ws", axum::routing::get(websocket_handler))
}